mars/learn/proxima/simple_index/builder.py

```diff
@@ -0,0 +1,292 @@
+        if op.topk is not None:
+            nsplit = cls._get_atleast_topk_nsplit(nsplit, op.topk)
+
+        if tensor.chunk_shape[1] > 1:
+            tensor = tensor.rechunk({0: nsplit, 1: tensor.shape[1]})._inplace_tile()
+        else:
+            tensor = tensor.rechunk({0: nsplit})._inplace_tile()
+        pk = pk.rechunk({0: nsplit})._inplace_tile()
+
+        out_chunks = []
+        for chunk, pk_col_chunk in zip(tensor.chunks, pk.chunks):
+            chunk_op = op.copy().reset_key()
+            chunk_op._stage = OperandStage.map
+            out_chunk = chunk_op.new_chunk([chunk, pk_col_chunk],
+                                           index=pk_col_chunk.index)
+            out_chunks.append(out_chunk)
+
+        logger.warning(f"index chunks count: {len(out_chunks)} ")
+
+        params = out.params
+        params['chunks'] = out_chunks
+        params['nsplits'] = ((1,) * len(out_chunks),)
+        new_op = op.copy()
+        return new_op.new_tileables(op.inputs, kws=[params])
+
+    @classmethod
+    def _execute_map(cls, ctx, op: "ProximaBuilder"):
+        inp = ctx[op.tensor.key]
+        out = op.outputs[0]
+        pks = ctx[op.pk.key]
+        proxima_type = get_proxima_type(inp.dtype)
+
+        # holder
+        holder = proxima.IndexHolder(type=proxima_type,
+                                     dimension=op.dimension)
+        for pk, record in zip(pks, inp):
+            pk = pk.item() if hasattr(pk, 'item') else pk
+            holder.emplace(pk, record.copy())
+
+        # converter
+        meta = proxima.IndexMeta(proxima_type, dimension=op.dimension,
+                                 measure_name=op.distance_metric)
+        if op.index_converter is not None:
+            converter = proxima.IndexConverter(name=op.index_converter,
+                                               meta=meta, params=op.index_converter_params)
+            converter.train_and_transform(holder)
+            holder = converter.result()
+            meta = converter.meta()
+
+        # builder && dumper
+        builder = proxima.IndexBuilder(name=op.index_builder,
+                                       meta=meta,
+                                       params=op.index_builder_params)
+        builder = builder.train_and_build(holder)
+
+        path = tempfile.mkstemp(prefix='proxima-', suffix='.index')[1]
+        dumper = proxima.IndexDumper(name="FileDumper", path=path)
+        builder.dump(dumper)
+        dumper.close()
+
+        if op.index_path is None:
+            ctx[out.key] = path
+        else:
+            # write to external file system
+            fs = get_fs(op.index_path, op.storage_options)
+            filename = f'proxima-{out.index[0]}.index'
+            out_path = f'{op.index_path.rstrip("/")}/{filename}'
+            with fs.open(out_path, 'wb') as out_f:
+                with open(path, 'rb') as in_f:
+                    # 32M
+                    chunk_bytes = 32 * 1024 ** 2
```

Comment:
Can this size be configurable?

Reply:
No need to configure, only used to reduce potential memory cost.